### PR TITLE
feat: Add recipe to reproduce Tulu-3 DPO model

### PR DESCRIPTION
**Pull request by mrm-196**
# What does this PR do ?

Adds a recipe to easily reproduce [Tulu-3 DPO 8B](Llama-3.1-Tulu-3-8B-DPO) model using NeMo-RL

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```bash
uv run ./examples/run_dpo.py   --config ./examples/configs/recipes/llm/dpo-tulu3-8b.yaml
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

tulu3-8b-accuracy
tulu3-8b-loss

### CODE REVIEW

**Comment by mrm-196** [comment]
@parthchadha Sure, the plots are added.

**Review by parthchadha — DISMISSED** [review]
LGTM, @ashors1 can you review as well?

File: examples/configs/recipes/llm/dpo-tulu3-8b.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  max_total_sequence_length: 2048
+  optimizer:
+    name: "torch.optim.AdamW"
+    kwargs:
+      lr: 5.0e-7
+      weight_decay: 0.0
+  scheduler:
+    - name: "torch.optim.lr_scheduler.LinearLR"
+      kwargs:
+        start_factor: 1.0e-6
```
Comment (ashors1) [inline examples/configs/recipes/llm/dpo-tulu3-8b.yaml]: nit: I think this should be 0.1 to match the paper exactly, right?

**Comment by ashors1** [comment]
Generally LGTM. Do you have any intuition about why the train loss has such high variance?

File: nemo_rl/data/hf_datasets/tulu3.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    chosen_conversation = data["chosen"]
+    rejected_conversation = data["rejected"]
+
+    context = chosen_conversation[:-1]
+
+    # We assume that except last assistant response, all messages in
+    # chosen and rejected conversations are similar. Validating this...
+    assert json.dumps(context, ensure_ascii=False) == json.dumps(
+        context, ensure_ascii=False
+    ), (
```
Comment (ashors1) [inline nemo_rl/data/hf_datasets/tulu3.py]: Is there a typo here? This will always return True, right?

**Review by terrykong — CHANGES_REQUESTED** [review]
change is good with me, but checking internally if we need any special steps to allow references to this dataset. blocking the PR until we hear back

File: examples/configs/recipes/llm/dpo-tulu3-8b.yaml (hunk 1 above)
Comment (mrm-196) [reply examples/configs/recipes/llm/dpo-tulu3-8b.yaml]: I think this should be okay.

In the training code that Tulu-3 team open-sourced, here is [how the scheduler is instantiated](https://github.com/allenai/open-instruct/blob/63baaf44f227bf40306811a4a73e0edcca68dcb3/open_instruct/dpo_tune_cache.py#L746).
So the code relies on `get_scheduler` function from the `transformers` library. Now, looking at [how transformers handle this](https://github.com/huggingface/transformers/blob/2c0af41ce5c448f872f3222a75f56030fb2e5a88/src/transformers/optimization.py#L105C1-L108C102): It basically starts the learning rate from 0 and warms it up to the peak value. And then cools it down to 0 again.

So, I think using a very small `start_factor` should be ideal, as long as it does not introduce numerical instabilities.

File: nemo_rl/data/hf_datasets/tulu3.py (hunk 2 above)
Comment (mrm-196) [reply nemo_rl/data/hf_datasets/tulu3.py]: Thanks for the great catch! Fixed this.

File: examples/configs/recipes/llm/dpo-tulu3-8b.yaml (hunk 1 above)
Comment (ashors1) [reply examples/configs/recipes/llm/dpo-tulu3-8b.yaml]: Ah makes sense, thanks for the explanation!

**Comment by ashors1** [comment]
Sounds good, thanks for the explanation. One final question -- were there any evaluations you ran to verify that results matched with the original Tulu-3 DPO results? Or did you just use the loss and accuracy curves as a signal?

**Comment by mrm-196** [comment]
> Sounds good, thanks for the explanation. One final question -- were there any evaluations you ran to verify that results matched with the original Tulu-3 DPO results? Or did you just use the loss and accuracy curves as a signal?

Yes, I did run some of [these evaluations](https://github.com/allenai/open-instruct/tree/main/eval):

| Model Name | MATH (Flex EM) | IFEval (Prompt loose) | TruthfulQA (MC2) | MMLU (EM) | GSM8K (EM) |
|----------|----------|----------|----------|----------|----------|
| [Tulu-3 8B SFT Original (HF Checkpoint)](https://huggingface.co/allenai/Llama-3.1-Tulu-3-8B-SFT)    |     0.3134     |     0.682     |     0.47     |     0.6406     |     0.7414     |
| [Tulu-3 8B DPO Original (HF Checkpoint)](https://huggingface.co/allenai/Llama-3.1-Tulu-3-8B-DPO)    |     0.3938     |     0.7948     |     0.5473     |     0.6436     |     0.8431     |
| Tulu-3 8B DPO Reproduced (with NeMo-RL)    |     0.4262     |     0.8022     |     0.569     |     0.642     |     0.8158     |

**Note**: These evaluations were run only once, and they are not min/max/average of multiple training runs.

**Review by ashors1 — APPROVED** [review]

File: examples/configs/recipes/llm/dpo-tulu3-8b.yaml
Comment (terrykong) [inline examples/configs/recipes/llm/dpo-tulu3-8b.yaml:1]: do you mind changing the name of this yaml to conform with the others here? we usually name them like this:
```
<algo>-<model>-#n#g-<parallelism>-<opt:long><opt:v$N>.sh
```

but since we now have another dataset, maybe

```
<algo>-<model>-<opt:dataset>-#n#g-<parallelism>-<opt:long><opt:v$N>.sh
```

**Review by terrykong — COMMENTED** [review]
lgtm. i checked internally and we should be good

File: examples/configs/recipes/llm/dpo-tulu3-8b.yaml
Comment (mrm-196) [reply examples/configs/recipes/llm/dpo-tulu3-8b.yaml:1]: Sure, then I think the ideal name would be: `dpo-llama3.1-8b-tulu3-1n8g-fsdp2tp1.sh`

Tulu-3 models are finetuned Llama3.1 models and they share the same architecture. I think the naming above clearly shows the algorithm, model architecture, dataset, and resource/parallelization configs.

But I also see that repro recipes are following a different naming convention. For example: [grpo-deepscaler-1.5b-8K.yaml](https://github.com/NVIDIA-NeMo/RL/blob/main/examples/configs/grpo-deepscaler-1.5b-8K.yaml). And if we follow that naming convention, the current `dpo-tulu3-8b.yaml` should be good. What do you think @terrykong ?

**Comment by ashors1** [comment]
@mrm-196 these changes look good to me, but in order for our CI to pass, we need to add a test for the Tulu-3 config that was added. This requires updating [tests/test_suites/llm](https://github.com/NVIDIA-NeMo/RL/tree/main/tests/test_suites/llm) to add a script called `dpo-llama3.1-8b-tulu3-1n8g-fsdp2tp1.sh` which runs the convergence test (you can use [dpo-llama3.1-8b-instruct-4n8g-megatron.sh](https://github.com/NVIDIA-NeMo/RL/blob/main/tests/test_suites/llm/dpo-llama3.1-8b-instruct-4n8g-megatron.sh) as a reference), and then adding `tests/test_suites/llm/dpo-llama3.1-8b-tulu3-1n8g-fsdp2tp1.sh` to [our set of release tests](https://github.com/NVIDIA-NeMo/RL/blob/main/tests/test_suites/release.txt#L27). Is that something you'd be able to do?

Also, I think there's a typo in the renamed config file. It's currently named "examples/configs/recipes/llm/dpo-llama3.1-8b-tulu3-1n8g-fsdp2tp1.**sh**" but should be "examples/configs/recipes/llm/dpo-llama3.1-8b-tulu3-1n8g-fsdp2tp1.**yaml**"

**Review by terrykong — APPROVED** [review]

**Review by terrykong — DISMISSED** [review]

**Review by terrykong — DISMISSED** [review]

**Comment by mrm-196** [comment]
Results of the test for this recipe:

Screenshot 2025-08-15 at 2 23 51 PM

**Review by terrykong — APPROVED** [review]

**Review by terrykong — APPROVED** [review]
